### ISSUE TITLE
Dataloader and Recursively nested query

### ISSUE THREAD

**Issue by danleydmello** [graphql-dotnet/graphql-dotnet#1250]
This question is related to  #1248. I have this recursively nested query:
```
query {
 ClassA {
     Prop2 {
       Prop2{
          Prop1
        }
      }
  }
}
```

which is based on the nested schema:
```
ClassA {
int Prop1,
ClassA Prop2
}
```

I am using dataloader for optimization. For each complex node, it creates loaderKey which is set to the name of the Field. In above query, as you see Prop2 field is nested twice, in which case loaderkey will be set to the same field, which is wrong. I want loaderkey to be different for each Prop2 field at a different level. Is there a way to uniquely identify, in the resolver, each field?

Thanks
Danley

**Comment by joemcbride** [comment]
I'm fairly certain the key is used for the _type_ of data, not a specific field instance.  Because the field is of the same type, that is fine.  When the field gets executed it would be getting different arguments and those arguments should be passed to the dataloader.

**Comment by danleydmello** [comment]
@joemcbride , the loaderKey is uniquely identify each loader (https://graphql-dotnet.github.io/docs/guides/dataloader). In the case of recursion, it's hard to identify the nested nodes of the same name. That's what I want to know if there is a way to identify each node in the query?

**Comment by joemcbride** [comment]
Why do you need to uniquely identify the node?  You should be able to just pass the identity of the object you're trying to resolve to the loader.


```csharp
Field<UserType, User>()
            .Name("User")
            .ResolveAsync(context =>
            {
            	// Get or add a batch loader with the key "GetUsersById"
            	// The loader will call GetUsersByIdAsync for each batch of keys
                var loader = accessor.Context.GetOrAddBatchLoader<int, User>("GetUsersById", users.GetUsersByIdAsync);

                // Add this UserId to the pending keys to fetch
                // The task will complete once the GetUsersByIdAsync() returns with the batched results
                return loader.LoadAsync(context.Source.UserId);
            });
```

In this example, every time this `User` field is is resolved, regardless of what level in the node hierarchy it is resolved at, it should always be able to provide a `UserId` value.

**Comment by danleydmello** [comment]
We need to need to uniquely identify the node for at least 2 reasons in our case:
1) We fetch and return the fields that are queried by the user and not the whole object.
  For e.g. in below query first node of "Prop2" should return instances of its child "Prop2" and the second node of "Prop2" should return scalar field Prop1.
```
query {
 ClassA {
     Prop2 {
       Prop2{
          Prop1
        }
      }
  }
}
```
Because these 2 nodes share same loader, the second node of "Prop2" will return "Prop1" as null because it uses the loader from node 1 where there is no "Prop1" field.

2. We have a filtering mechanism for each complex node. In below query you can do this:
```
query {
 ClassA {
     Prop2 **(prop1: 3)** {
       Prop2 **(prop1:5)**{
          Prop1
        }
      }
  }
}
```
Same problem as above, since both these nodes share the same loader the filter of the first node will be applied to the second one.

This is why at resolver time I need to uniquely identify the node so that I create a unique loader for each.

Thanks
Danley

**Comment by danleydmello** [comment]
@joemcbride , any suggestion on the above comment?

**Comment by Shane32** [comment]
@danleydmello I have code with similar needs.  I used a ValueTuple and combined the `IResolveFieldContext` and the key both in a single variable to use as the dataloader key.  Then the dataloader would perform its own grouping analysis when it executes, executing the groups independently if needed.  In my case, I grouped it on the AST field object, so that it would be grouped separately depending on where in the original query the field was requested.

Let me know if you have further questions.
